### ISSUE TITLE
CFP: Completing Mutual Authentication + Stable Maturity

### ISSUE THREAD

**Issue by tgraf** [cilium/cilium#28986]
This is a follow-on issue to track the road to feature completeness and stable maturity following the work that was merged in Cilium 1.14 (https://github.com/cilium/cilium/issues/22215)

**Current maturity state:** beta
**Security model completeness:** Not complete yet

This meta issue tracks progress on the [Next-Generation Mutual Authentication with Cilium Service Mesh](https://isovalent.com/blog/post/2022-05-03-servicemesh-security/) implementation.

The [CFP](https://docs.google.com/document/d/1Do9_ICj4ZokSiWqXYdb7tQvhR--Rz2GyHJ2Xk6NDhQU/edit?usp=sharing) describes how we are planning to implement this, why we've chosen SPIFFE and SPIRE over something less complex, and so on. Here, we track actually getting the work done.

### Work performed in 1.14

The issue https://github.com/cilium/cilium/issues/22215 lists the detailed work done. 1.14 delivered an initial implementation of mutual authentication with per identity authentication. Per-identity authentication was implemented first to validate the feasibility of caching authentication handshakes for performance reasons, it will unlikely be the default model of operation.

### Work scheduled for 1.15 and beyond

  - [ ] Support Wireguard to encrypt & authenticate the identity used for mutual authentication
  - [ ] Per connection authentication
  - [ ] Pruning of ipcache based on authentication effectiveness
  - [ ] Detailed documentation of the security model
  - [ ] Conduct penetration test of implementation and document findings
  - [x] #23808
  - [ ] #25730 (this actually requires a more nuanced discussion, but this feature will _not_ default to on for 1.14).
  - [x] #26480 
  - [x] #28357 
  - [x] #24617
  - [x] #31430 
  - [ ] Make decision on whether to mark the feature stable

**Comment by cccsss01** [comment]
With the purchase of isovalent how is this road map looking?

**Comment by youngnick** [comment]
Thanks for the ping @cccsss01, we're working on the plan here and hope to have some more updates in the next couple of weeks.

I can say that we don't anticipate that the acquisition of Isovalent will have any impact at all on this roadmap though.

**Comment by youngnick** [comment]
I've just opened cilium/design-cfps#27, outlining the changes we are going to make to both the existing mutual auth method, and  to the connection-based mutual auth method that we've started work on internally.

There are two main actions:
- Update existing documentation to make clear that Wireguard (preferable) or at least tunneling mode (less secure because it's not wire-encrypted) are requirements for turning on mutual auth in any mode. This includes updating existing mutual auth docs and install instructions to make this easy for folks.
- Build out new connection-based mutual auth method, as designed in cilium/design-cfps#27. An early PoC of this work is up an running but needs some cleanup before being made publically available.

I'll update this issue as further action is taken here (I'll be opening PRs to update the docs in the next couple of days).

**Comment by cccsss01** [comment]
rip

**Comment by cccsss01** [comment]
@youngnick how are things looking?

**Comment by youngnick** [comment]
We're still in the planning phase for now, we're working on getting some resources prioritized here to make some more progress though.

**Comment by cccsss01** [comment]
Thank you for the update, was really hoping to get rid of istio :(

**Comment by youngnick** [comment]
Not stale, still coming.

**Comment by an-luu-ts** [comment]
any update?

**Comment by youngnick** [comment]
No updates currently, other things are currently ahead in the queue.

**Comment by ugoenyioha** [comment]
When foks?

**Comment by mikn** [comment]
My understanding of WG and the purpose of this task seems to put them as strictly incompatible, sort of as noted here: https://github.com/cilium/cilium/issues/26480

Are you considering alternative encryption methods of the pod to pod comms with this enabled? There is [riptides](https://riptides.io/) which seems to take the approach of a separate kernel module to avoid another user space process. I am unsure whether you can get away with anything less if you do not want to run sidecars?

**Comment by husira** [comment]
any update regarding this topic?

**Comment by manu-ns** [comment]
I guess the update is here https://github.com/cilium/cilium/releases/tag/v1.19.0
> 🚇 Ztunnel Beta: You can enroll namespaces into Ztunnel, which enables TCP connections between workloads to be transparently encrypted and authenticated. (https://github.com/cilium/cilium/pull/42766, https://github.com/cilium/cilium/pull/42819, https://github.com/cilium/cilium/pull/43227 and others, @ldelossa, @rgo3, @nddq)
> 👥 Mutual Authentication: The out-of-band [Mutual Authentication](https://docs.cilium.io/en/v1.19.0/network/servicemesh/mutual-authentication/mutual-authentication/) feature is now disabled by default, pending community feedback. If you have a requirement for mTLS, consider trying the new Ztunnel integration. (https://github.com/cilium/cilium/pull/42665, @christarazi)

**Comment by youngnick** [comment]
Yes, we received a lot of pushback from users who felt that calling this "mTLS" was problematic. Combine that with the lack of further community contributions here, and the general cooling off of this space in favor of ztunnel, along with having community members interested in ztunnel integration, and Mutual Auth does not look like very viable.

Interested in any community feedback about the feature, however. Please let us know what you  think here.

**Comment by youngnick** [comment]
With the deprecation of Mutual Auth (under #47132), this issue is obsolete. Closing to make outstanding issues easier to track.
